### PR TITLE
add rule jsx-sort-props to order props alphabetically, optionally by type

### PR DESCRIPTION
**Pull request by SteveByerly**
Addresses #8 

Also see similar update in tslint:
https://github.com/palantir/tslint/pull/2831

CHANGELOG.md entry:

[new-rule] `jsx-sort-props`: Sort props alphabetically and optionally by type (spread, shorthand, longhand)

### CODE REVIEW

File: src/rules/jsxSortPropsRule.ts [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+
+const OPTION_IGNORE_CASE = "ignore-case";
+
+const OPTION_ORDER_SPREAD = "spread";
+const OPTION_ORDER_SHORTHAND = "shorthand";
+const OPTION_ORDER_LONGHAND = "longhand";
+type ORDER_OPTION = "spread" | "shorthand" | "longhand";
+
+enum KEY_ORDER {
+  OPTION_ORDER_SPREAD,
```
Comment (adidahiya) [inline src/rules/jsxSortPropsRule.ts]: consistent 4-space indentation please

File: src/rules/jsxSortPropsRule.ts [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+            Individual strings that control general functionality.
+            Possible settings are:
+
+            * \`"${OPTION_IGNORE_CASE}"\`: Ignore case when comparing props.
+
+            An array of strings that control ordering of props by type.
+            Possible settings are:
+
+            * \`"${OPTION_ORDER_SPREAD}"\`: Spread assignments \`{...props}\`.
+            * \`"${OPTION_ORDER_SHORTHAND}"\`: Shorthand properties \`{ isError, isValid }\`.
```
Comment (adidahiya) [inline src/rules/jsxSortPropsRule.ts]: shorthand conflicts with `jsx-boolean-value: true`, which we enable by default... https://github.com/palantir/tslint-react/blob/master/tslint-react.json#L5

**Comment by SteveByerly** [comment]
Will address based on the outcome of [my reply](https://github.com/palantir/tslint/pull/2831#issuecomment-304950414).

Spread would no longer be a part of this change, and other than adding alpha-sorting, it would simply allow for ensuring shorthand come first and we can alpha-sort within those groupings.

**Comment by SteveByerly** [comment]
Updated based on the same approach I took here: https://github.com/palantir/tslint/pull/2831#issuecomment-310855009

**Comment by david0178418** [comment]
Looking forward to this feature.  @adidahiya , did the modifications made by @SteveByerly address your concerns?

**Comment by rootulp** [comment]
ping @palantir/tslint

**Comment by johnwiseheart** [comment]
Not sure what the issue is with the builds here, but I'd guess @SteveByerly may need to push once more to have the tests run (and hopefully pass). If they pass, I'd be happy to merge this in.

**Comment by bradleyayers** [comment]
Any updates on this one @johnwiseheart ?

**Comment by jdolinski1** [comment]
Will it be possible to automatically fix unsorted props with tslint --fix?

**Comment by bradleyayers** [comment]
That would be fantastic!

**Comment by adidahiya** [comment]
closing as stale, sorry it took so long to review @SteveByerly. let us know if it's ok for someone to take take your code and make a new PR.

**Comment by daviddelusenet** [comment]
Will this ever be picked up again?

**Pull request closed without merging**